packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx

```diff
@@ -1,34 +1,19 @@
-import { Component, ReactNode, ComponentType } from 'react';
-import { connect } from 'react-redux';
-import debounce from 'lodash/debounce';
-import compose from 'recompose/compose';
-import { createSelector } from 'reselect';
-import get from 'lodash/get';
-import isEqual from 'lodash/isEqual';
-import difference from 'lodash/difference';
-
 import {
-    crudGetMany as crudGetManyAction,
-    crudGetMatching as crudGetMatchingAction,
-} from '../../actions/dataActions';
-import {
-    getPossibleReferences,
-    getPossibleReferenceValues,
-    getReferenceResource,
-} from '../../reducer';
-import { getStatusForArrayInput as getDataStatus } from './referenceDataStatus';
-import withTranslate from '../../i18n/translate';
-import { Record, Sort, Translate, Pagination, Dispatch } from '../../types';
-import { MatchingReferencesError } from './types';
+    ComponentType,
+    FunctionComponent,
+    ReactElement,
+    useCallback,
+} from 'react';
+import debounce from 'lodash/debounce';
 
-const defaultReferenceSource = (resource: string, source: string) =>
-    `${resource}@${source}`;
+import { Record, Sort, Pagination } from '../../types';
+import useReferenceArrayInputController from './useReferenceArrayInputController';
 
 interface ChildrenFuncParams {
     choices: Record[];
     error?: string;
+    loaded: boolean;
     loading: boolean;
-    onChange: (value: any) => void;
```

Comment:
I still don't understand why the new controller doesn't return this one.

Reply:
> Because `onChange` override is handled at the input level. See the `useInput` hook